packages/nimble-components/src/table/specs/table-column-specs/table-column-mapping.md

```diff
@@ -0,0 +1,299 @@
+
+### Background
+
+[Icon column type issue](https://github.com/ni/nimble/issues/1013)
+
+[Boolean column type issue](https://github.com/ni/nimble/issues/1103)
+
+### Features
+
+-   Supported input:
+    -   string
+    -   number
+    -   boolean
+-   Supported output:
+    -   Text
+    -   Icon
+    -   Spinner
+    -   (empty)
+
+### Non-goals
+
+-   Non-Nimble icon support
+-   Arbitrary icon colors
+-   Hyperlink icons
+-   Mixed text and icons
+-   Non-icon, non-spinner Nimble components
+
+---
+
+## Design
+
+Below is an example of how these elements would be used within a `nimble-table`:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-icon field-name="status">
+        Status
+        <nimble-mapping-icon key="fail" icon="nimble-icon-xmark" severity="error" label="Failed"></nimble-mapping-icon>
+        <nimble-mapping-icon key="error" icon="nimble-icon-xmark" severity="error" label="Errored"></nimble-mapping-icon>
+        <nimble-mapping-icon key="pass" icon="nimble-icon-check" severity="success" label="Passed"></nimble-mapping-icon>
+        <nimble-mapping-spinner key="running" label="Running..."></nimble-mapping-spinner>
+    </nimble-table-column-icon>
+    <nimble-table-column-mapping field-name="errorCode" data-type="number">
+        Error Code
+        <nimble-mapping-text key="1" label="A bad thing happened"></nimble-mapping-text>
+        <nimble-mapping-text key="2" label="A worse thing happened"></nimble-mapping-text>
+        <nimble-mapping-text key="3" label="A terrible thing happened"></nimble-mapping-text>
+    </nimble-table-column-mapping>
+    <nimble-table-column-icon field-name="archived" data-type="boolean">
+        Archived
+        <nimble-mapping-icon key="true" icon="nimble-icon-database" label="Archived"></nimble-mapping-icon>
+    </nimble-table-column-icon>
+</nimble-table>
+```
+
+Note that the `key` attribute values are always given as strings. In the case of boolean or number mappings, this value is parsed to the appropriate type, so that it can properly be compared to the values from the row records.
+
+When none of the given mappings match the record value for a cell, that cell will be empty. Alternatively, we could provide support for marking a mapping with a `default` attribute that would cause it to be chosen when no other values matched. This would be equivalent to the `placeholder` configuration we provide on `nimble-table-column-text` and `nimble-table-column-anchor`. I don't think we need to support this initially.
+
+If multiple mappings in a column have the same key, an error will be thrown.
+
+If an invalid `icon` value is passed to `nimble-mapping-icon`, an error will be thrown.
```

Comment:
I think we should be more specific about the validation being done. Specifically, some questions I have are:
- What makes an icon valid vs invalid? Is a valid icon any element that has `Icon` as its base class?
- Why throw an error? We've generally made the statement that setting an attribute on an element shouldn't throw an error because there isn't a good way to catch that error. Also, I could see the validation code being async, [similar to the column filtering code in the table](https://github.com/ni/nimble/blob/486a75bd2be240c186a627a6ceb93f433b14a478/packages/nimble-components/src/table/index.ts#L565), because we don't know if all custom elements have been defined yet. Should this valid/invalid state really bubble up to the table's `validity` object in some way?

Reply:
> What makes an icon valid vs invalid? Is a valid icon any element that has Icon as its base class?

That seems like an accurate way to do it. I'll update the spec.

> Why throw an error? ... Should this valid/invalid state really bubble up to the table's validity object in some way?

I welcome some guidance here. Naively, it sounds like I should set some flags in the table's validity object instead of throwing errors. Are there any reasons I shouldn't?

Reply:
Seems like there is a lot of validation needed for this column type compared to others so far it may be worthwhile to figure out something now. I'll say anecdotally having that added to the table made it a lot easier to reason about the behavior of error conditions.

API-wise, just brainstorming, I think something like:
```ts
abstract class Validator {
    abstract isValid()
}

class TableValidator extends Validator {
   private invalidColumnConfiguration // true if one or more invalid columns
}

class IconColumnValidator extends Validator {
}

class TableColumn {
    abstract readonly columnValidity: Validator;
}
```
Let's the table not need to know about specific column details and columns can determine there own properties. Just one idea off-the-top.

Reply:
I agree that correct configuration is pretty fundamental to getting these new column types to behave properly, so we shouldn't defer figuring validation out. However, I'm almost always in favor of making smaller PRs -- just because we say we shouldn't defer the work doesn't mean it has to all be part of one giant PR.

Another note, is that after thinking through this a bit more, I've realized that there are really two different types of validation for a column - validation that the data records match the types the columns are expecting and validation of the configuration of the column & any children it has. The issue I linked in my comment was thinking about the first type of validation. The majority of the validation you need to do here is the second type of validation, which doesn't need access to the data records and can be determined entirely by the column with no additional information (i.e. doesn't need to be a function that gets called on the column). I haven't fully thought through @rajsite's example API, but I think we can potentially leverage `@observable` properties on the column to help communicate a column's validity back up to the table in a way that isn't possible when the table's data is required to validate a column.

Reply:
I'm still unsure how much detail we want to surface with regards to invalid column configuration. We could do what Milan had in his suggestion and have just a single flag for any column being invalid, but it would be nice if we could give more specifics. Should we let the table validator store specific validation error strings that it collects from individual columns? Should columns have separate validity objects that the client has to read from? (That seems overly inconvenient.)

> we can potentially leverage @observable properties on the column...

Is there something significant about having `@observable` properties over just validity-checking functions?

Reply:
>> we can potentially leverage `@observable` properties on the column...

> Is there something significant about having `@observable` properties over just validity-checking functions

Yes, there is a fairly significant architectural difference between `@observable` properties vs validity-checking functions -- who is responsible for knowing that the validity may have changed & needs to be rechecked. With `@observable` properties, the column can own all of its own logic for what could make it invalid (e.g. specific attributes changing, child items changing, etc) and tell the table when the state has changed. With validity-checking functions, the table somehow has to know that the validity of a given column might have changed and so the function needs to be called again. Therefore, I think with `@observable` properties it would be pretty straightforward to have the column own its own validity and not leak implementation details into the table. It's possible this can be done with a function, but I think it wouldn't be as clear and/or clean.

Reply:
Sure, I was just thinking that with the current scheme of the validation object only being updated when the `validate()` function is called, column validation functions could just be called when all the existing validation functions get called. But `@observable`s facilitate updating validation at other times, which may be useful.

Reply:
I've updated the table column HLD with the plan for column validation.